python/lsst/ip/isr/isrTaskLSST.py

```diff
@@ -1346,77 +1403,190 @@ def run(self, ccdExposure, *, dnlLUT=None, bias=None, deferredChargeCalib=None,
         if self.config.doBrighterFatter:
+            if bfKernel is None:
+                raise RuntimeError("doBrighterFatter is True not no bfKernel provided.")
             self.compareCameraKeywords(exposureMetadata, bfKernel, "brighter-fatter")
         if self.config.doFlat:
+            if flat is None:
+                raise RuntimeError("doFlat is True but not flat provided.")
             self.compareCameraKeywords(exposureMetadata, flat, "flat")
 
+        # FIXME: Make sure that if linearity is done then it is matched
+        # with the right PTC.
+
         # We keep track of units: start in ADU.
-        exposureMetadata["LSST ISR UNITS"] = "ADU"
+        exposureMetadata["LSST ISR UNITS"] = "adu"
 
-        # First we convert the exposure to floating point values.
-        if self.config.doConvertIntToFloat:
-            self.log.info("Converting exposure to floating point values.")
-            ccdExposure = self.convertIntToFloat(ccdExposure)
+        # First we convert the exposure to floating point values
+        # (if necessary).
+        self.log.debug("Converting exposure to floating point values.")
+        ccdExposure = self.convertIntToFloat(ccdExposure)
 
         # Then we mark which amplifiers are completely bad from defects.
         badAmpDict = self.maskFullDefectAmplifiers(ccdExposure, detector, defects)
 
         # Now we go through ISR steps.
+
+        # Differential non-linearity correction.
+        # Units: ADU
         if self.config.doDiffNonLinearCorrection:
             self.diffNonLinearCorrection(ccdExposure, dnlLUT)
 
+        # Dither the integer counts.
+        # Input units: integerized ADU
+        # Output units: floating-point ADU
+        self.ditherCounts(ccdExposure, overscanDetectorConfig)
+
+        nominalPtcUsed = False
+        if self.config.doBootstrap or ptc is None:
+            self.log.info(
+                "Configured using doBootstrap=True; using nominal gain of %.3f.",
+                self.config.nominalGain,
+            )
+            nominalPtcUsed = True
+            ptc = PhotonTransferCurveDataset([amp.getName() for amp in detector], "NOMINAL_PTC", 1)
+            for amp in detector:
+                ptc.gain[amp.getName()] = self.config.nominalGain
+                ptc.noise[amp.getName()] = 0.0
+
+        exposureMetadata["LSST ISR NOMINAL PTC USED"] = nominalPtcUsed
+
+        gains = ptc.gain
+
+        # And check if we have configured gains to override. This is
+        # also a warning, since it should not be typical usage.
+        for amp in detector:
+            if not math.isnan(gain := overscanDetectorConfig.getOverscanAmpConfig(amp).gain):
+                gains[amp.getName()] = gain
+                self.log.warning(
+                    "Overriding gain for amp %s with configured value of %.3f.",
+                    amp.getName(),
+                    gain,
+                )
+
+        # Serial overscan correction.
+        # Units: ADU
```

Comment:
If the header keyword is lower case, can we make these guides lower case as well?  I'm close to suggesting an isrFunction to add the header keywords so we can enforce the set of possible values it can take, and correct "ADU"->"adu" and "electrons"->"electron".

Reply:
Interesting point!  I can add that function and fix the units in the comments.  (The lower case is because this is what astropy units wants).